src/Storages/MergeTree/IntersectionsIndexes.h

```diff
@@ -0,0 +1,206 @@
+            }
+        }
+
+        return intersected_before ? IntersectionResult::EXACTLY_ONE_INTERSECTION : IntersectionResult::NO_INTERSECTION;
+    }
+
+    using OrderedSegments = std::set<PartToRead>;
+    OrderedSegments segments;
+};
+
+struct HalfIntervals
+{
+    static HalfIntervals initializeWithEntireSpace()
+    {
+        auto left_inf = std::numeric_limits<decltype(MarkRange::begin)>::min();
+        auto right_inf = std::numeric_limits<decltype(MarkRange::end)>::max();
+        return HalfIntervals{{{left_inf, right_inf}}};
+    }
+
+    static HalfIntervals initializeFromMarkRanges(MarkRanges ranges)
+    {
+        OrderedRanges new_intervals;
+        for (const auto & range : ranges)
+            new_intervals.insert(range);
+
+        return HalfIntervals{std::move(new_intervals)};
+    }
+
+    MarkRanges convertToMarkRangesFinal()
+    {
+        MarkRanges result;
+        std::move(intervals.begin(), intervals.end(), std::back_inserter(result));
+        return result;
+    }
+
+    HalfIntervals & intersect(const HalfIntervals & rhs)
+    {
+        /**
+         * first   [   ) [   ) [   ) [  ) [  )
+         * second    [       ) [ ) [   )  [    )
+         */
+        OrderedRanges intersected;
+
+        const auto & first_intervals = intervals;
+        auto first = first_intervals.begin();
+        const auto & second_intervals = rhs.intervals;
+        auto second = second_intervals.begin();
+
+        while (first != first_intervals.end() && second != second_intervals.end())
+        {
+            auto are_intersect = [](auto & x, auto & y)
+            {
+                if ((x->begin <= y->begin) && (y->begin < x->end))
+                    return true;
+                if ((y->begin <= x->begin) && (x->begin < y->end))
+                    return true;
+                return false;
+            };
+
+            if (are_intersect(first, second))
+            {
+                intersected.insert(MarkRange{
+                    std::max(second->begin, first->begin),
+                    std::min(second->end, first->end)
+                });
+            }
+
+            if (first->end <= second->end)
+                ++first;
+            else
+                ++second;
+        }
+
+        std::swap(intersected, intervals);
+
+        return *this;
+    }
+
+    HalfIntervals & negate()
+    {
+        auto left_inf = std::numeric_limits<decltype(MarkRange::begin)>::min();
+        auto right_inf = std::numeric_limits<decltype(MarkRange::end)>::max();
+        OrderedRanges new_ranges;
+
+        /// Possibly add (-inf; begin)
+        if (!intervals.empty())
+            if (auto begin = intervals.begin()->begin; begin != left_inf)
+                new_ranges.insert(MarkRange{left_inf, begin});
+
+        auto prev = intervals.begin();
+        for (auto it = std::next(intervals.begin()); it != intervals.end(); ++it)
+        {
+            if (prev->end != it->begin)
```

Comment:
I would rather carefully check that this is not possible (looks like it's true).

Reply:
It is possible. Imagine [0, 1) [1, 2). This condition is needed not to include [1,1) in the answer.